### PR TITLE
Helptickets: Log text tickets

### PR DESCRIPTION
**Pull request by mia-pi-git**
Useful feature. We currently log to modlog but that's extremely inconvenient.

### CODE REVIEW

File: server/chat-plugins/helptickets.ts [hunk 1]
```diff
@@ -404,6 +413,47 @@ export class HelpTicket extends Rooms.RoomGame {
...
+			text: ticket.text,
+			resolved: ticket.resolved,
+			meta: ticket.meta,
+			created: ticket.created,
+			userid: ticket.userid,
+			type: ticket.type,
+			claimed: ticket.claimed,
+		};
+		const date = Chat.toTimestamp(new Date()).split(' ')[0];
+		void FS(`logs/tickets/${date.slice(0, -3)}.jsonl`).append(JSON.stringify(entry) + '\n');
```
Comment (AnnikaCodes) [inline server/chat-plugins/helptickets.ts:427]: I think these should probably be in directories by month.

Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:427]: Why a directory? That's a lot of tiny files that are pretty hard to sort apart.

Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:427]: I'm not opposed, it just seems more difficult.

Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:427]: Ripgrep seems more efficient than trying to readdir potentially many files out of one dir at once.

Comment (AnnikaCodes) [reply server/chat-plugins/helptickets.ts:435]: I think it's better than ripgrepping over a directory containing hundreds if not thousands of files. I'm not sure why you're concerned about "trying to readdir potentially many files out of one dir at once" when my suggestion _reduces_ the number of files in each directory.

Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:435]: Discussed on PS and resolved. It's one file a month, whereas this comment was made thinking it was one a day.

Comment (AnnikaCodes) [reply server/chat-plugins/helptickets.ts:435]: Apparently I misunderstood and this adds one file per month, which is fine.

File: server/chat-plugins/helptickets.ts [hunk 2]
```diff
@@ -412,6 +421,47 @@ export class HelpTicket extends Rooms.RoomGame {
...
+		if (await checkRipgrepAvailability()) {
+			const args = [`-e`, `userid":"${userid}`, '--no-filename'];
+			const lines = await ProcessManager.exec([
+				`rg`, `${__dirname}/../../logs/tickets/${date ? `${date}.jsonl` : ''}`, ...args,
+			]);
+			for (const line of lines.stdout.split('\n')) {
+				if (line.trim()) results.push(JSON.parse(line));
+			}
+		} else {
+			if (!date) throw new Chat.ErrorMessage(`Specify a date.`);
```
Comment (AnnikaCodes) [inline server/chat-plugins/helptickets.ts:448]:
```suggestion
			if (!date) throw new Chat.ErrorMessage(`Specify a month.`);
```

File: server/chat-plugins/helptickets.ts [hunk 3]
```diff
@@ -412,6 +421,47 @@ export class HelpTicket extends Rooms.RoomGame {
...
+			const path = FS(`logs/tickets/${date}.jsonl`);
+			if (!path.existsSync()) {
+				throw new Chat.ErrorMessage(`There are no logs for that date.`);
```
Comment (AnnikaCodes) [inline server/chat-plugins/helptickets.ts:451]:
```suggestion
				throw new Chat.ErrorMessage(`There are no logs for the month "${date}".`);
```

File: server/chat-plugins/helptickets.ts [hunk 4]
```diff
@@ -2169,6 +2288,7 @@ export const commands: Chat.ChatCommands = {
 		`/helpticket ignore - Ignore notifications for unclaimed help tickets. Requires: % @ &`,
 		`/helpticket unignore - Stop ignoring notifications for help tickets. Requires: % @ &`,
 		`/helpticket delete [user] - Deletes a user's ticket. Requires: &`,
+		`/helpticket logs [userid][, date] - View logs of the [userid]'s text tickets. Requires: % @ &`,
```
Comment (AnnikaCodes) [inline server/chat-plugins/helptickets.ts:2291]:
```suggestion
		`/helpticket logs [userid][, month] - View logs of the [userid]'s text tickets. Requires: % @ &`,
```

File: server/chat-plugins/helptickets.ts [hunk 5]
```diff
@@ -2159,6 +2265,19 @@ export const commands: Chat.ChatCommands = {
...
+		logs(target, room, user) {
+			this.checkCan('lock');
+			const [targetString, dateString] = Utils.splitFirst(target, ',').map(i => i.trim());
+			const id = toID(targetString);
+			if (!id) return this.errorReply(`Specify a userid.`);
+			return this.parse(`/j view-help-logs-${id}${dateString ? `--${dateString}` : ''}`);
+		},
+		logshelp: [
+			`/helpticket logs [userid][, date] - View logs of the [userid]'s text tickets. `,
+			`If a [date] is given, searches only that month.`,
```
Comment (AnnikaCodes) [inline server/chat-plugins/helptickets.ts:2277]:
```suggestion
			`/helpticket logs [userid][, month] - View logs of the [userid]'s text tickets. `,
			`If a [month] is given, searches only that month.`,
```

**Review by AnnikaCodes — APPROVED** [review]
This looks good to me! I will leave it for a day or two in case anyone has any comments.

File: server/chat-plugins/helptickets.ts (hunk 5 above)
Comment (AnnikaCodes) [reply server/chat-plugins/helptickets.ts:2277]: [17:48] mia: you can commit that review to my PR btw
[17:49] mia: I'm on phone so it's hard

**Pull request merged**